### ISSUE TITLE
The Error propagation for Swapping Action is not correct

### ISSUE THREAD

**Issue by Naphann** [sfc-aqua/quisp#343]
# Describe the Bug

Since QuISP works on error basis model, using the measurement result as eigenvalue could result in incorrect behavior.

https://github.com/sfc-aqua/quisp/blob/06447115c79dab8da8255aee0e7e0668c69b6254/quisp/rules/actions/SwappingAction.cc#L63-L86

In QuISP, we cannot know the correct Pauli frame correction to be applied since we do not work on density matrices. We can see that the code is incorrect when we look at the perfect simulation case. Even if the two Bell pairs to be swapped are in Phi+ state, the result of Bell measurement is still equally uniform for all of the 4 Bell bases. The measurement result mid-operation does not tell us anything about what's going on. What measurement in each basis should do is to inform about how the propagation of the error should be done. 

The snippet below 

https://github.com/sfc-aqua/quisp/blob/e74de6b082a9a1934124e0268d9280610a590198/quisp/modules/QNIC/StationaryQubit/StationaryQubit.cc#L245-L251

Are only useful at the end of simulation when we want to calculate the fidelity of the result states. 

## Suggestion for the Fix

Introduce new methods for measurement in the 3 bases, to be used during the RuleSet operation which only propagates errors and does not return anything (make the method signature be `void`). Also, rename the old measurement method to be specifically used for tomography/fidelity calculation only.

**Comment by Naphann** [comment]
We had discussion about this regarding stabilizer states. If all the states we're working with are all stabilizer states, the local measurement of these states will be completely mixed meaning that the fix should 


### The fix to measurement operation

1. Randomly select the measurement output either (1 or -1) and correctly flag this result (if needed) for the affected entangled parties
2. Propagate the error from the measured qubit to the entangled parties. Similar to what is shown in [kaaki's thesis](https://arxiv.org/abs/1908.10758) page 67.

**Comment by ngym** [comment]
You are correct. Since pauli error simulator does not know the actual state of qubits, it cannot provide measurement outcomes, unless extrapolating information about the state as context. 

For stabilizer measurement, the eigenvalue "should "be always +1, unless any error happens. Therefore measurement of pauli error simulator does not cause problems. This is why pauli error simulator is often used for error correction research.

From the point of view of compatibility, adding a context as entanglement swapping to quantum states is more desirable and keeping RuleSet genuine than modifying RuleSet for this pauli error simulation-specific problem. Cannot we add such context to quantum states?

**Comment by Naphann** [comment]
Since we are not doing the stabilizer measurement but single qubit Pauli measurement (X, Y, Z) if what I'm understanding from Fred's last night is correct the outcome of stabilizer state (in the context of QuISP only graph state) the eigenvalues of the measurement are always 50-50 for the two eigenvalues. 

The additional tracking that we need is outside of the RuleSet scope, so the RuleSet is still genuine. The additional tracking we need will be on the simulation layer (or currently called the "GOD" layer) and this tracking is needed since we must know the ideal state (so we can propagate the error). It's just haven't been added properly in QuISP since we are only at the Bell pair level and no multiparty states yet. 

So in summary, yes, the additional context isn't for RuleSet. RuleSet doesn't need to change. 
Although the original problems in this issue is that the RuleSet is wrong...

**Comment by Naphann** [comment]
fixed via PR #351
